File: java/src/ggz/cards/client/Player.java
/*
 * Copyright (C) 2006  Helg Bredow
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package ggz.cards.client;

import ggz.cards.common.Bid;
import ggz.cards.common.Card;
import ggz.common.SeatType;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * A "player" as seen by the client is really just a seat at the table, complete
 * with a hand and everything. It's not necessary that a player actually be
 * sitting at that seat - there may just be a pile of cards.
 * 
 * @note We are always player 0. The server will deal with all conversions.
 *       Contains all information about a seat at the table.
 */
public class Player {

    /** ggz seating assignment info */
    SeatType status;

    /** player's name */
    String name;

    /** GGZ seat number (or -1) */
    int ggzseat;

    /** For games that support teams (or partnerships). */
    int team;

    /**
     * @brief Player's card on the table.
     * 
     * This variable is maintained by the core client-common code, and should
     * not be modified by the GUI table code. It will always hold the card known
     * to be on the table for the player, i.e. basically the same information
     * that the server knows (although with a bit of lag). Note that this need
     * not correspond with what is actually drawn by the GUI.
     */
    Card tableCard;

    /** player's hand */
    ArrayList hand;

    /** Bid player made for the current hand. */
    Bid bid;

    /** true if the player is currently bidding. */
    boolean bidding;

    /** true if the player is currently playing a card. */
    boolean playing;

    /** The number of tricks this player has won this hand. */
    int tricks;

    public String getName() {
        return this.name;
    }

    public List getHand() {
        return Collections.unmodifiableList(this.hand);
    }

    public Card getTableCard() {
        return this.tableCard;
    }

    public SeatType getSeatType() {
        return this.status;
    }

    public Bid getBid() {
        return this.bid;
    }

    public int getTeam() {
        return this.team;
    }

    public boolean isBidding() {
        return this.bidding;
    }

    public boolean isPlaying() {
        return this.playing;
    }
    
    public int getTricks() {
        return this.tricks;
    }

    /**
     * The seat number as recorded by GGZ. Game seat numbers are from the
     * perspective of each client, this give the "absolute" number and is the
     * number that should be used for reseat requests.
     * 
     * @return
     */
    public int getGGZSeatNum() {
        return this.ggzseat;
    }
}

File: ggz/ggz/player.c
/*
 * File: player.c
 * Author: Brent Hendricks
 * Project: GGZ Client
 * Date: 6/5/00
 *
 * This fils contains functions for handling players
 *
 * Copyright (C) 1998 Brent Hendricks.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */


#include <glib.h>

#include <err_func.h>
#include <player.h>
#include <chat.h>


/* List  of players in current room */
static GList* players;

/* Local functions for manipulating player list */
static void player_free_name(gpointer, gpointer);
static void player_list_print(void);
static void player_print(gpointer, gpointer);
static gint player_match_name(gconstpointer, gconstpointer);


void player_list_clear(void)
{
	g_list_foreach(players, player_free_name, NULL); 
	g_list_free(players);
	players = NULL;
}


void player_list_add(gchar* name, gint table, gint color)
{
	Player* player;

	chat_print(CHAT_COLOR_SERVER, "-->", name);

	dbg_msg("Adding %s to player list", name);
	
	player = g_malloc(sizeof(Player));
	player->name = g_strdup(name);
	player->table = table;
	player->chat_color = color;

	players = g_list_append(players, (gpointer)player);
	player_list_print();
}


void player_list_remove(gchar* name)
{
	GList* node;

	chat_print(CHAT_COLOR_SERVER, "<--", name);

	dbg_msg("Removing %s from player list", name);
	
	node = g_list_find_custom(players, (gpointer)name, player_match_name);
	/* name not found */
	if (!node)
		return;

	players = g_list_remove_link(players, node);
	g_list_foreach(node, player_free_name, NULL); 
	g_list_free(node);
	player_list_print();
}


void player_list_update(gchar* name, gint table, gint color)
{
	GList* node;
	Player* player;

	dbg_msg("Updating player list info for %s", name);

	node = g_list_find_custom(players, (gpointer)name, player_match_name);
	/* name not found */
	if (!node)
		return;

	/* Update information */
	player = (Player*)node->data;
	player->table = table;
	player->chat_color = color;

	player_list_print();
}


void player_list_iterate(GFunc func)
{
	g_list_foreach(players, func, NULL);
}


static void player_free_name(gpointer player, gpointer data)
{
	if (((Player*)player)->name)
		g_free(((Player*)player)->name);
}


static void player_list_print(void)
{
	g_list_foreach(players, player_print, NULL); 
}


static void player_print(gpointer player, gpointer data)
{
	dbg_msg("Player name: %s", ((Player*)player)->name);
	dbg_msg("Player table: %d", ((Player*)player)->table);
	dbg_msg("Player color: %d", ((Player*)player)->chat_color);
}


static gint player_match_name(gconstpointer player, gconstpointer name)
{
	return strcmp(((Player*)player)->name, name);
}

